### ISSUE TITLE
Change GET favorites/ to return list of post JSON objects

### ISSUE THREAD

**Issue by samuelrey** [SoftwareEngineeringDaily/software-engineering-daily-api#95]
At the moment, the endpoint returns a list of favorite objects that look like
```JSON
[{
	"_id": "5a0e0c6ad6b0a2002ad17dcd",
	"userId": "5a021a15454aef0028432711",
	"postId": "5a0d62069e36ee0004c6402d",
	"__v": 0,
	"active": false
}, {
	"_id": "5a0e2534d6b0a2002ad17dd0",
	"userId": "5a021a15454aef0028432711",
	"postId": "5a0c107c41bcad00045a7467",
	"__v": 0,
	"active": true
}, {
	"_id": "5a0e2538d6b0a2002ad17dd1",
	"userId": "5a021a15454aef0028432711",
	"postId": "5a0abf1762969e00041c640f",
	"__v": 0,
	"active": true
}]
```
But the posts themselves displayed, not the favorites.

**Comment by joshmorel** [comment]
This definitely makes more sense to me. I'm working on something related see #83 and will work on this next once that is done (unless someone wants to take this on).

I assume you would only want those that have active=true for the authenticated user, correct?

**Comment by samuelrey** [comment]
Read through the issue you linked. Relating the users to posts via bookmarks is definitely a larger problem than the one I've brought up here. I'm not familiar with mongo, but it seems like you've got a grasp on the situation.

@jasonify and I spoke yesterday & he agreed to take on this issue, but thanks for the help 😄

As for the active=true bit, I'm actually not sure what the value is meant to represent. If I understand you correctly though, then I do want only the posts bookmarked by the authenticated user.

**Comment by jasonify** [comment]
@samuelrey , let's have @joshmorel  handle this one

**Comment by joshmorel** [comment]
I'm starting to work on this and I'm wondering if grouping these routes returning user-specific lists make more sense to group under user. Note I'll use the term bookmarks/bookmarked as this is label we're moving towards.

So for example:

* `/api/users/bookmarked`: Returns list of bookmarked posts for authenticated user. Later we can add relatedLinks, or query parameters to return the types of items that were bookmarked.
* `/api/users/{userId}/bookmarked`: Returns list of bookmarked posts for specific user. 

Alternately for current user we can use `api/users/me/bookmarked` to be really explicit.

You can see this is the approach used in stack exchange: https://api.stackexchange.com/docs

**Comment by samuelrey** [comment]
The most RESTful design would be to use `api/users/{userid}/bookmarks` for all users. See https://cloud.google.com/apis/design/

**Comment by joshmorel** [comment]
I'll implement this request api/users/{userId}/bookmarked. We can decide what we want to do with GET api/favorites

I still wonder if a "me" shortcut might still be beneficial for client developer convenience. @jasonify do you have a preference?

**Comment by jasonify** [comment]
Yeah I prefer that personally. I am using `api/users/me` for the api already here: 
https://github.com/SoftwareEngineeringDaily/software-engineering-daily-api/blob/develop/server/routes/user.route.js

**Comment by justinctlam** [comment]
Going back to the original comment, GET favorites/ is returning a list of items with postId for favorites/bookmarks.

Couple of questions:

*  Is GET favorites going to be renamed to GET bookmarks?
    * Same question for POST favorite and POST unfavorite
*  Can we get the entire post data returned from the GET API?
     * This is because we don't want to do individual network fetches to get the post id again, better to get it all at once with paging. We can't assume we have all the bookmarked posts when we do a regular GET posts/ call.

**Comment by justinctlam** [comment]
Also, I don't quite understand why we have active=false posts in the response data. I think it's better to just have active=true response data as mentioned above. This will make it clean and clear that the items returned from the list are actually favorited.

**Comment by crablar** [comment]
@justinctlam Yes--we should rename favorites -> bookmarks everywhere. And I think it's fine to get the entire post data.

As far as "active", I don't know what that field means. @jasonify ?

**Comment by jasonify** [comment]
For the votes the "active" field is supposed to be similar to how you have a "deleted" field so you don't actually delete anything. I assume it's the same here?  I do think it's somewhat confusing .

**Comment by justinctlam** [comment]
Great! Thank you! Please let me know when these APIs get updated.

**Comment by joshmorel** [comment]
For all new routes I've been using bookmark(s/ed). But for existing ones, for example to bookmark a post, it is still favorite. I can create synonym routes for these in separate PR - related to issue #99 

Eventually we should remove the "favorite" ones but definitely not right away.

Let me know if this is a worthwhile next thing to do after #117 .

**Comment by justinctlam** [comment]
@joshmorel Have you already submitted the new Android changes? If not, we should update the APIs now before release and change our code asap. Otherwise, maybe the API team can leave the favorites but deprecate the APIs in a month or two for your changes to propagate.

**Comment by joshmorel** [comment]
I haven't been working on Android so I'm quite sure where that's at.

This was my thought, for example for bookmarking a post, we'd support both of these:

* `/api/post/{postId}/favorite`
* `/api/post/{postId}/bookmark`

We can also add a deprecated flag in swagger for the first. But this way all the different clients can switch their routes to use bookmark when it is convenient. I don't think a big bang switch for the back-end and ALL the clients is a good idea. 

Once we know all clients are migrated in production we can then remote the existing favorite routes.

I'll create a PR for issue #99  and get feedback.

**Comment by joshmorel** [comment]
@justinctlam - this issue is resolved right? I figure we should close it. But also FYI PR #131 now provides all the previous favorite routes as bookmark as well.

**Comment by justinctlam** [comment]
Yes, feel free to close. Thank you!
